errai-common/src/main/java/org/jboss/errai/common/client/dom/elemental2/Elemental2DomUtil.java

```diff
@@ -0,0 +1,99 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.jboss.errai.common.client.dom.elemental2;
+
+import com.google.gwt.user.client.ui.RootPanel;
+import com.google.gwt.user.client.ui.Widget;
+import elemental2.dom.HTMLElement;
+import elemental2.dom.Node;
+import jsinterop.base.Js;
+
+/**
+ * Provides utility methods for interacting with the DOM.
+ * @author Guilherme Carreiro <ggomes@redhat.com>
+ */
+public class Elemental2DomUtil {
+
+    /**
+     * Detaches all element children from a node.
+     * @param node Must not be null.
+     * @return True iff any element children were detached by this call.
+     */
+    public static boolean removeAllElementChildren(final Node node) {
+
+        final boolean hadChildren = node.lastChild != null;
+
+        while (node.lastChild != null) {
+            node.removeChild(node.lastChild);
+        }
+
+        return hadChildren;
+    }
+
+    /**
+     * Appends the underlying {@link HTMLElement} of a {@link Widget} to another {@link HTMLElement},
+     * in a way that does not break GWT Widget events.
+     * @param parent The parent element that is appended to. Must not be null.
+     * @param child The child Widget, whose underlying HTML element will be
+     * appended to the parent. Must not be null.
+     */
+    public static void appendWidgetToElement(final HTMLElement parent, final Widget child) {
+
+        if (child.isAttached()) {
+            child.removeFromParent();
+        }
+
+        RootPanel.detachOnWindowClose(child);
+
+        HTMLElement newChild = asHTMLElement(child.getElement());
+        parent.appendChild(newChild);
+
+        attach(child);
+    }
+
+    /**
+     * Converts from {@link com.google.gwt.dom.client.Element} to {@link HTMLElement}.
+     * @param gwtElement The deprecated GWT Element
+     * @return The GWT Elemental2 HTMLElement
+     */
+    public static HTMLElement asHTMLElement(final com.google.gwt.dom.client.Element gwtElement) {
+        return Js.cast(gwtElement);
+    }
+
+    /**
+     * Converts from {@link org.jboss.errai.common.client.dom.HTMLElement} to {@link HTMLElement}.
+     * @param htmlElement The deprecated Errai HTMLElement
+     * @return The GWT Elemental2 HTMLElement
+     */
+    public static HTMLElement asHTMLElement(final org.jboss.errai.common.client.dom.HTMLElement htmlElement) {
+        return Js.cast(htmlElement);
+    }
+
+    private static void attach(final Widget child) {
+        final Elemental2Widget widget = Js.uncheckedCast(child);
```

Comment:
To be honest, this makes me uneasy. As far as I know, `Js.cast` and `Js.uncheckedCast` are meant for casting between native JavaScript types, not GWT/Java types. How have you tested this? Does it work in a production compiled showcase?

My concern here is that in certain compilations (like a production compile with full optimizations) this cast might succeed but the method call to onAttach might fail because the underlying object is not the correct type.

Reply:
@mbarkley Yeah, these casts are a little bit nasty. But, unfortunately, I see them as the solution with least impact that we have to solve this problem right now.

I tested that on a production compiled showcase, and the components work as expected. The following Gif shows a component (Accordion) that depends of the `Elemental2DomUtil`:

[gif - interagindo com accordion] 1

Since the signature of the `attach` method is:
```java
void attach(final Widget child)`
````
...and the `Widget` has a protected method called `onAttach`, I don't think that we're going to see a `child` without the `onAttach` method being called.

Am I missing something?

Thanks.

Reply:
So I gave this a try in the Errai tutorial, and there is an error calling attach when production compiled. I suspect this screenshot was from one of the webapps that is configured to use draft compile, even when you do a `mvn clean package` (without the "full" profile).

[Here](https://github.com/mbarkley/errai-tutorial/tree/unchecked-cast-repro) is a reproducer in the errai-tutorial. Just run `mvn clean package`, deploy the app to WildFly, open the app, and look in the console for the log message "Something went terribly wrong".

The fix, I think, is to change the attach method back to JSNI. This is an instance where JSNI is the only solution until we are ready to migrate to GWT3 and have a GWT3-specific replacement.

Reply:
@mbarkley I recorded the Gif using a war from `kie-wb-distributions` (master) on WildFly 10.1.0.

But, I'm going to change it to JSNI for now :-) Thanks for the tip.